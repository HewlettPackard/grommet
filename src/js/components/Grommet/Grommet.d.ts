import * as React from 'react';
import Grommet from '../index';

export declare namespace GrommetTypes {
  type Direction = 'rtl' | 'ltr';
}

export interface GrommetProps extends Grommet.Props {
  dir: GrommetTypes.Direction;
  theme: {};
}

export class Grommet extends React.Component<GrommetProps, undefined> { }

export default Grommet;
