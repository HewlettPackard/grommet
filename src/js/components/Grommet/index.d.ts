import * as React from "react";

export interface GrommetProps {
  full?: boolean;
  theme?: object;
}

declare const Grommet: React.StatelessComponent<GrommetProps>;

export { Grommet };
