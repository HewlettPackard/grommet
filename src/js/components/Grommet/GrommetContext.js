import React from 'react';

// Track which components mounted onto the DOM are coming from
// Grommet (see FocusedContainer.js)
export const GrommetContext = React.createContext(undefined);
