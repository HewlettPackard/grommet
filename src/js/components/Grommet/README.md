## Grommet
This is the top level Grommet container.

[![](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/github/grommet/grommet-site?initialpath=grommet&amp;module=%2Fscreens%2FGrommet.js)
## Usage

```javascript
import { Grommet } from 'grommet';
<Grommet>...</Grommet>
```

## Properties

**dir**

Whether text should be rendered right to left or not. Defaults to
      inherit from the document context.

```
rtl
ltr
```

**theme**

Custom styles for Grommet app component.

```
object
```
  