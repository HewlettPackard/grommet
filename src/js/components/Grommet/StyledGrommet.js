import styled, { css } from 'styled-components';

import { baseStyle } from '../../utils';

const fullStyle = css`
  width: 100vw;
  height: 100vh;
`;

const StyledGrommet = styled.div`
  ${baseStyle}
  ${props => props.full && fullStyle}
  ${props => props.theme.global.font.face}
`;

export default StyledGrommet.extend`
  ${props => props.theme.grommet.extend}
`;
