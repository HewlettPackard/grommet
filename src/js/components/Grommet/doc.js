import { describe, PropTypes } from 'react-desc';

import { getAvailableAtBadge } from '../../utils';

export default (Grommet) => {
  const DocumentedGrommet = describe(Grommet)
    .availableAt(getAvailableAtBadge('Grommet'))
    .description('This is the top level Grommet container.')
    .usage(
      `import { Grommet } from 'grommet';
<Grommet>...</Grommet>`
    );

  DocumentedGrommet.propTypes = {
    dir: PropTypes.oneOf(['rtl', 'ltr']).description(
      `Whether text should be rendered right to left or not. Defaults to
      inherit from the document context.`
    ),
    theme: PropTypes.object.description('Custom styles for Grommet app component.'),
  };

  return DocumentedGrommet;
};
