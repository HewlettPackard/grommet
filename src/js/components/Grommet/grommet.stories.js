import React from 'react';
import { storiesOf } from '@storybook/react';

import { Grommet, Anchor, Box } from 'grommet';
import { Add } from 'grommet-icons';

const customTheme = {
  global: {
    colors: {
      custom: '#cc6633',
    },
  },
};


const Themed = () => (
  <Grommet theme={customTheme}>
    <Box pad='medium'>
      <Anchor icon={<Add />} label='Add' color='custom' />
    </Box>
  </Grommet>
);

storiesOf('Grommet', module)
  .add('Theme', () => <Themed />);
