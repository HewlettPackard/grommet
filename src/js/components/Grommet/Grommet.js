import React, { forwardRef, useEffect, useMemo, useState } from 'react';
import { createGlobalStyle } from 'styled-components';

import {
  ContainerTargetContext,
  ResponsiveContext,
  ThemeContext,
} from '../../contexts';

import {
  deepMerge,
  backgroundIsDark,
  getBreakpoint,
  getDeviceBreakpoint,
  normalizeColor,
} from '../../utils';
import { base as baseTheme } from '../../themes';
import { StyledGrommet } from './StyledGrommet';

const FullGlobalStyle = createGlobalStyle`
  body { margin: 0; }
`;

const deviceResponsive = (userAgent, theme) => {
  // log('--deviceResponsive', userAgent, theme);
  /*
   * Regexes provided for mobile and tablet detection are meant to replace
   * a full-featured specific library due to contributing a considerable size
   * into the bundle.
   *
   * User agents found https://deviceatlas.com/blog/list-of-user-agent-strings
   */
  if (userAgent) {
    if (/(tablet|ipad|playbook|silk)|(android(?!.*mobile))/i.test(userAgent)) {
      return getDeviceBreakpoint('tablet', theme);
    }
    if (/Mobile|iPhone|Android/.test(userAgent)) {
      return getDeviceBreakpoint('phone', theme);
    }
    return getDeviceBreakpoint('computer', theme);
  }
  return undefined;
};

const Grommet = forwardRef((props, ref) => {
  const {
    children,
    full,
    containerTarget = typeof document === 'object' ? document.body : undefined,
    theme: themeProp,
    ...rest
  } = props;

  const { background, dir, themeMode, userAgent } = props;

  const [stateResponsive, setResponsive] = useState();

  const theme = useMemo(() => {
    const nextTheme = deepMerge(baseTheme, themeProp || {});

    const {
      colors: { background: themeBackground },
    } = nextTheme.global;

    nextTheme.dark = (themeMode || nextTheme.defaultMode) === 'dark';
    const color = normalizeColor(background || themeBackground, nextTheme);
    nextTheme.dark = backgroundIsDark(color, nextTheme);
    nextTheme.baseBackground = background || themeBackground;
    // This allows DataTable to intelligently set the background of a pinned
    // header or footer.
    nextTheme.background = nextTheme.baseBackground;

    if (dir) {
      nextTheme.dir = dir;
    }

    return nextTheme;
  }, [background, dir, themeMode, themeProp]);

  useEffect(() => {
    const onResize = () => {
      setResponsive(getBreakpoint(window.innerWidth, theme));
    };
    window.addEventListener('resize', onResize);
    onResize();
    return () => {
      window.removeEventListener('resize', onResize);
    };
  }, [theme]);

  const responsive =
    stateResponsive ||
    deviceResponsive(userAgent, theme) ||
    theme.global.deviceBreakpoints.tablet;

  return (
    <ThemeContext.Provider value={theme}>
      <ResponsiveContext.Provider value={responsive}>
        <ContainerTargetContext.Provider value={containerTarget}>
          <StyledGrommet full={full} {...rest} ref={ref}>
            {children}
          </StyledGrommet>
          {full && <FullGlobalStyle />}
        </ContainerTargetContext.Provider>
      </ResponsiveContext.Provider>
    </ThemeContext.Provider>
  );
});

Grommet.displayName = 'Grommet';

let GrommetDoc;
if (process.env.NODE_ENV !== 'production') {
  // eslint-disable-next-line global-require
  GrommetDoc = require('./doc').doc(Grommet);
}
const GrommetWrapper = GrommetDoc || Grommet;

export { GrommetWrapper as Grommet };
