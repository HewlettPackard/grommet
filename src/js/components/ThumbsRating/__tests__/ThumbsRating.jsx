import React from 'react';
import { render } from '@testing-library/react';
import { axe } from 'jest-axe';

import 'jest-styled-components';
import 'jest-axe/extend-expect';
import 'regenerator-runtime/runtime';
import '@testing-library/jest-dom';

import { Grommet } from '../../Grommet';
import { ThumbsRating } from '..';

describe('ThumbsRating', () => {
  test('ThumbsRating with string options', () => {
    const { container } = render(
      <Grommet>
        <ThumbsRating name="test" options={['positive', 'negative']} />
      </Grommet>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  test('ThumbsRating with color star green', () => {
    const { container } = render(
      <Grommet>
        <ThumbsRating name="test" color="green" />
      </Grommet>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  test('ThumbsRating with color outline and fill', () => {
    const { container } = render(
      <Grommet>
        <ThumbsRating name="test" color={{ outline: 'green', fill: 'pink' }} />
      </Grommet>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  test('ThumbsRating with defaultValue', () => {
    const { container } = render(
      <Grommet>
        <ThumbsRating name="test" options={['one', 'two']} defaultValue="one" />
      </Grommet>,
    );

    expect(container).toMatchSnapshot();
  });
});
