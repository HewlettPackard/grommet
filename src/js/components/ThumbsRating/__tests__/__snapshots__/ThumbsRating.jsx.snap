// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThumbsRating StarRating should have no accessibility violations 1`] = `
.c5 {
  display: inline-block;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  width: 24px;
  height: 24px;
  fill: #666666;
  stroke: #666666;
}

.c5 g {
  fill: inherit;
  stroke: inherit;
}

.c5 *:not([stroke])[fill="none"] {
  stroke-width: 0;
}

.c5 *[stroke*="#"],
.c5 *[STROKE*="#"] {
  stroke: inherit;
  fill: none;
}

.c5 *[fill-rule],
.c5 *[FILL-RULE],
.c5 *[fill*="#"],
.c5 *[FILL*="#"] {
  fill: inherit;
  stroke: none;
}

.c0 {
  font-size: 18px;
  line-height: 24px;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
}

.c6 {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  -webkit-align-self: stretch;
  -ms-flex-item-align: stretch;
  align-self: stretch;
  width: 12px;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  cursor: pointer;
}

.c2:hover input:not([disabled]) + div,
.c2:hover input:not([disabled]) + span {
  border-color: #000000;
}

.c4 {
  opacity: 0;
  -moz-appearance: none;
  width: 0;
  height: 0;
  margin: 0;
  cursor: pointer;
}

@media only screen and (max-width:768px) {
  .c6 {
    width: 6px;
  }
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
      role="radiogroup"
    >
      <label
        class="c2"
        for="like"
      >
        <div
          class="c3 "
        >
          <input
            class="c4"
            id="like"
            name="test"
            tabindex="0"
            type="radio"
            value="like"
          />
          <svg
            aria-label="Like"
            class="c5"
            viewBox="0 0 24 24"
          >
            <path
              d="M1 23h19c2 0 3-1 3-3V10h-7V4c0-2-1-3-3-3h-2s-.016 6-.016 7.326C10.984 9.652 10 11 8 11H1v12zm5 0V11"
              fill="none"
              stroke="#000"
              stroke-width="2"
            />
          </svg>
        </div>
      </label>
      <div
        class="c6"
      />
      <label
        class="c2"
        for="dislike"
      >
        <div
          class="c3 "
        >
          <input
            class="c4"
            id="dislike"
            name="test"
            tabindex="-1"
            type="radio"
            value="dislike"
          />
          <svg
            aria-label="Dislike"
            class="c5"
            viewBox="0 0 24 24"
          >
            <path
              d="M23 1H4C2 1 1 2 1 4v10h7v6c0 2 1 3 3 3h2s.016-6 .016-7.326C13.016 14.348 14 13 16 13h7V1zm-5 0v12"
              fill="none"
              stroke="#000"
              stroke-width="2"
            />
          </svg>
        </div>
      </label>
    </div>
  </div>
</div>
`;

exports[`ThumbsRating ThumbsRating with color outline and fill 1`] = `
.c5 {
  display: inline-block;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  width: 24px;
  height: 24px;
  fill: green;
  stroke: green;
}

.c5 g {
  fill: inherit;
  stroke: inherit;
}

.c5 *:not([stroke])[fill="none"] {
  stroke-width: 0;
}

.c5 *[stroke*="#"],
.c5 *[STROKE*="#"] {
  stroke: inherit;
  fill: none;
}

.c5 *[fill-rule],
.c5 *[FILL-RULE],
.c5 *[fill*="#"],
.c5 *[FILL*="#"] {
  fill: inherit;
  stroke: none;
}

.c0 {
  font-size: 18px;
  line-height: 24px;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
}

.c6 {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  -webkit-align-self: stretch;
  -ms-flex-item-align: stretch;
  align-self: stretch;
  width: 12px;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  cursor: pointer;
}

.c2:hover input:not([disabled]) + div,
.c2:hover input:not([disabled]) + span {
  border-color: #000000;
}

.c4 {
  opacity: 0;
  -moz-appearance: none;
  width: 0;
  height: 0;
  margin: 0;
  cursor: pointer;
}

@media only screen and (max-width:768px) {
  .c6 {
    width: 6px;
  }
}

<div
  class="c0"
>
  <div
    class="c1"
    role="radiogroup"
  >
    <label
      class="c2"
      for="like"
    >
      <div
        class="c3 "
      >
        <input
          class="c4"
          id="like"
          name="test"
          tabindex="0"
          type="radio"
          value="like"
        />
        <svg
          aria-label="Like"
          class="c5"
          viewBox="0 0 24 24"
        >
          <path
            d="M1 23h19c2 0 3-1 3-3V10h-7V4c0-2-1-3-3-3h-2s-.016 6-.016 7.326C10.984 9.652 10 11 8 11H1v12zm5 0V11"
            fill="none"
            stroke="#000"
            stroke-width="2"
          />
        </svg>
      </div>
    </label>
    <div
      class="c6"
    />
    <label
      class="c2"
      for="dislike"
    >
      <div
        class="c3 "
      >
        <input
          class="c4"
          id="dislike"
          name="test"
          tabindex="-1"
          type="radio"
          value="dislike"
        />
        <svg
          aria-label="Dislike"
          class="c5"
          viewBox="0 0 24 24"
        >
          <path
            d="M23 1H4C2 1 1 2 1 4v10h7v6c0 2 1 3 3 3h2s.016-6 .016-7.326C13.016 14.348 14 13 16 13h7V1zm-5 0v12"
            fill="none"
            stroke="#000"
            stroke-width="2"
          />
        </svg>
      </div>
    </label>
  </div>
</div>
`;

exports[`ThumbsRating ThumbsRating with color star green 1`] = `
.c5 {
  display: inline-block;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  width: 24px;
  height: 24px;
  fill: green;
  stroke: green;
}

.c5 g {
  fill: inherit;
  stroke: inherit;
}

.c5 *:not([stroke])[fill="none"] {
  stroke-width: 0;
}

.c5 *[stroke*="#"],
.c5 *[STROKE*="#"] {
  stroke: inherit;
  fill: none;
}

.c5 *[fill-rule],
.c5 *[FILL-RULE],
.c5 *[fill*="#"],
.c5 *[FILL*="#"] {
  fill: inherit;
  stroke: none;
}

.c0 {
  font-size: 18px;
  line-height: 24px;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
}

.c6 {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  -webkit-align-self: stretch;
  -ms-flex-item-align: stretch;
  align-self: stretch;
  width: 12px;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  cursor: pointer;
}

.c2:hover input:not([disabled]) + div,
.c2:hover input:not([disabled]) + span {
  border-color: #000000;
}

.c4 {
  opacity: 0;
  -moz-appearance: none;
  width: 0;
  height: 0;
  margin: 0;
  cursor: pointer;
}

@media only screen and (max-width:768px) {
  .c6 {
    width: 6px;
  }
}

<div
  class="c0"
>
  <div
    class="c1"
    role="radiogroup"
  >
    <label
      class="c2"
      for="like"
    >
      <div
        class="c3 "
      >
        <input
          class="c4"
          id="like"
          name="test"
          tabindex="0"
          type="radio"
          value="like"
        />
        <svg
          aria-label="Like"
          class="c5"
          viewBox="0 0 24 24"
        >
          <path
            d="M1 23h19c2 0 3-1 3-3V10h-7V4c0-2-1-3-3-3h-2s-.016 6-.016 7.326C10.984 9.652 10 11 8 11H1v12zm5 0V11"
            fill="none"
            stroke="#000"
            stroke-width="2"
          />
        </svg>
      </div>
    </label>
    <div
      class="c6"
    />
    <label
      class="c2"
      for="dislike"
    >
      <div
        class="c3 "
      >
        <input
          class="c4"
          id="dislike"
          name="test"
          tabindex="-1"
          type="radio"
          value="dislike"
        />
        <svg
          aria-label="Dislike"
          class="c5"
          viewBox="0 0 24 24"
        >
          <path
            d="M23 1H4C2 1 1 2 1 4v10h7v6c0 2 1 3 3 3h2s.016-6 .016-7.326C13.016 14.348 14 13 16 13h7V1zm-5 0v12"
            fill="none"
            stroke="#000"
            stroke-width="2"
          />
        </svg>
      </div>
    </label>
  </div>
</div>
`;
